examples/direct_fidelity_estimation.py

```diff
@@ -276,6 +277,35 @@ def _estimate_pauli_traces_general(qubits: List[cirq.Qid],
     return pauli_traces
 
 
+def _estimate_std_devs_clifford(fidelity: float, n: int) -> Tuple[float, float]:
+    """
+    Estimates the standard deviation of the measurement for Clifford circuits.
+
+    Args:
+        fidelity: The measured fidelity
+        n: the number of measurements
+
+    Returns:
+        The standard deviation (estimated from the fidelity and a maximum bound
+        on the variance regardless of what the true fidelity is)
+    """
+
+    # We use the Bhatia Davis inequality to estimate the variance of the
+    # fidelity. This gives that:
+    # Var[\hat{F}] <= (1 - F) . F / N
+    # StdDev[\hat{F}] <= \sqrt{(1 - F) . F / N}
+    #
+    # By further using the fact that 0 <= F <= 1 we get:
+    # StdDev[\hat{F}] <= \frac{1}{2 \sqrt{N}}
+
+    fidelity_bounded = max(0.0, min(1.0, fidelity))
```

Comment:
is clipping the right thing to do here? or should we throw an exception if the input is outside of the bounds it's supposed to be in

Reply:
I think it's because we use the *estimated* fidelity, which can be outside the range [0, 1], even if we know that the true fidelity cannot be outside of it, right?

If I comment out the range breaking, I can see some unit tests failing like here:
```
   def test_calling_main():
        dfe.main(n_measured_operators=3, samples_per_term=0)
>       dfe.main(n_measured_operators=3, samples_per_term=10)

examples/direct_fidelity_estimation_test.py:180: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
examples/direct_fidelity_estimation.py:498: in main
    estimated_fidelity, _ = direct_fidelity_estimation(
examples/direct_fidelity_estimation.py:448: in direct_fidelity_estimation
    std_dev_estimate, std_dev_bound = _estimate_std_devs_clifford(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

fidelity = -0.06666666666666667, n = 3
```

I looked at the code + math a little bit more, but with some noise, it's not impossible that we flip the sign right? I'm still not sure at all and there could be a bug in the existing code of course. Also, I'm not sure it's a good idea to use the estimated fidelity to estimate its variance. I don't have the math to support it, but it seems like it'd be reasonable if the estimate is good enough.

For these two reasons, I'm not resolving the conversation. Thanks for the very careful review!

Reply:
0.06 seems like a pretty big deviation, although for 3 operators and 10 samples per term, you'd also expect a high variance. I can think of three options and I'll defer to you what you think is best:

1. Add an additional `tolerance` parameter with a reasonably small default so an error can be thrown if someone inputs something blatantly wrong (like a fidelity of 23) and tests can expand the tolerance when statistics are known to be bad
2. Do something fancier like throw an exception if the deviation from [0,1] exceeds the computed standard deviation :)
3. Make an argument that since this is a private helper function we know that the input will be a fidelity so the chance of blatant abuse is lessened or would be caught by the rest of the fidelity tests